File: Src/Engine/Editor/GUI/ComboBox.cpp

#include "ComboBox.h"

#include "ContextMenu/ContextMenu.h"
#include "ContextMenu/ContextMenuButton.h"
#include "Runtime/Core/Types/Collections/ListExtensions.h"
#include "Runtime/Core/Types/Collections/Sorting.h"
#include "Runtime/Graphics/GraphicWindow.h"
#include "Runtime/Render/2D/Render2D.h"
#include "Runtime/UI/GUI/Style.h"
#include "Runtime/UI/GUI/WindowRootControl.h"
#include "Runtime/UI/GUI/Brushes/IBrush.h"
#include "Runtime/UI/GUI/Brushes/SpriteBrush.h"

namespace SE::Editor
{
	ComboBox::ComboBox() : ComboBox(0, 0)
	{
	}
	ComboBox::ComboBox(float x, float y, float width) : Control(x, y, width, DefaultHeight)
	{
		MaximumItemsInViewCount = 20;

		Style* style = Style::Current;
		Font = FontReference(style->FontMedium);
		TextColor = style->Foreground;
		BackgroundColor = style->BackgroundNormal;
		BackgroundColorHighlighted = BackgroundColor;
		BackgroundColorSelected = BackgroundColor;
		BorderColor = style->BorderNormal;
		BorderColorHighlighted = style->BorderSelected;
		BorderColorSelected = BorderColorHighlighted;
		ArrowImage = New<SpriteBrush>(style->ArrowDown);
		ArrowColor = style->Foreground * 0.6f;
		ArrowColorSelected = style->BackgroundSelected;
		ArrowColorHighlighted = style->Foreground;
	}

	void ComboBox::ClearItems()
	{
		SelectedIndex = -1;
		items.Clear();
	}

	void ComboBox::AddItem(String& item)
	{
		items.Add(item);
	}

	void ComboBox::AddItems(List<String>& items)
	{
		items.Add(items);
	}

	void ComboBox::SetItems(List<String>& items)
	{
		SelectedIndex = -1;
		items.Clear();
		items.Add(items);
	}

	bool ComboBox::IsSelected(String& item)
	{
		return IsSelected(items.Find(item));
	}

	bool ComboBox::IsSelected(int index)
	{
		return index != -1 && _selectedIndices.Contains(index);
	}
	
	void ComboBox::OnDestroy()
	{
		if (_popupMenu != nullptr)
		{
			_popupMenu->Hide();
			_popupMenu->Dispose();
			_popupMenu = nullptr;
		}

		if (IsDisposing)
			return;
		_selectedIndices.Clear();
		items.Clear();

		Control::OnDestroy();
	}
	
	void ComboBox::Draw()
	{
		// Cache data
		Rectangle clientRect = Rectangle(Float2::Zero, Size);
		float margin = clientRect.GetHeight() * 0.2f;
		float boxSize = clientRect.GetHeight() - margin * 2;
		bool isOpened = IsPopupOpened;
		bool enabled = EnabledInHierarchy();
		Color backgroundColor = BackgroundColor;
		Color borderColor = BorderColor;
		Color arrowColor = ArrowColor;
		if (!enabled)
		{
			backgroundColor *= 0.5f;
			arrowColor *= 0.7f;
		}
		else if (isOpened || _mouseDown)
		{
			backgroundColor = BackgroundColorSelected;
			borderColor = BorderColorSelected;
			arrowColor = ArrowColorSelected;
		}
		else if (IsMouseOver || IsNavFocused)
		{
			backgroundColor = BackgroundColorHighlighted;
			borderColor = BorderColorHighlighted;
			arrowColor = ArrowColorHighlighted;
		}

		// Background
		Render2D::FillRectangle(clientRect, backgroundColor);
		Render2D::DrawRectangle(clientRect.MakeExpanded(-2.0f), borderColor);

		// Check if has selected item
		if (_selectedIndices.Count() > 0)
		{
			String text = _selectedIndices.Count() == 1 ? (_selectedIndices[0] >= 0 && _selectedIndices[0] < items.Count() ? items[_selectedIndices[0]] : SE_TEXT("")) : SE_TEXT("Multiple Values");

			// Draw text of the selected item
			float textScale = Height / DefaultHeight;
			Rectangle textRect = Rectangle(margin, 0, clientRect.GetWidth() - boxSize - 2.0f * margin, clientRect.GetHeight());
			Render2D::PushClip(textRect);
			Color textColor = TextColor;
			Render2D::RenderText(Font.GetFont(), text, textRect, enabled ? textColor : textColor * 0.5f, TextAlignment::Near, TextAlignment::Center, TextWrapping::NoWrap, 1.0f, textScale);
			Render2D::PopClip();
		}

		// Arrow
		if (ArrowImage != nullptr)
		{
			ArrowImage->Draw(Rectangle(clientRect.GetWidth() - margin - boxSize, margin, boxSize, boxSize), arrowColor);
		}
	}
	
	void ComboBox::OnLostFocus()
	{
		Control::OnLostFocus();

		// Clear flags
		_mouseDown = false;
		_blockPopup = false;
	}
	
	void ComboBox::OnMouseLeave()
	{
		// Clear flags
		_mouseDown = false;
		_blockPopup = false;

		Control::OnMouseLeave();
	}
	
	bool ComboBox::OnMouseDown(Float2 location, MouseButton button)
	{
		if (button == MouseButton::Left)
		{
			_mouseDown = true;
			Focus();
			return true;
		}

		return Control::OnMouseDown(location, button);
	}
	
	bool ComboBox::OnMouseUp(Float2 location, MouseButton button)
	{
		if (_mouseDown && !_blockPopup)
		{
			_mouseDown = false;
			ShowPopup();
		}
		else
		{
			_blockPopup = false;
		}

		return true;
	}

	void ComboBox::OnSubmit()
	{
		Control::OnSubmit();

		ShowPopup();
	}

	void ComboBox::OnSelectedIndexChanged()
	{
		if (tooltips.Count() == items.Count())
		{
			TooltipText = _selectedIndices.Count() == 1 ? tooltips[_selectedIndices[0]] : nullptr;
		}
		SelectedIndexChanged(this);
	}

	void ComboBox::ShowPopup()
	{
		// Ensure to have valid menu
		if (_popupMenu == nullptr)
		{
			_popupMenu = OnCreatePopup();
			_popupMenu->MaximumItemsInViewCount = MaximumItemsInViewCount;

			// Bind events
			_popupMenu->VisibleChanged.Bind([this](Control* cm)
			{
				RootControl* win = Root;
				_blockPopup = win != nullptr && Rectangle(Float2::Zero, Size).Contains(PointFromWindow(win->GetMousePosition()));
				if (!_blockPopup)
					Focus();
			});
			_popupMenu->ButtonClicked.Bind([this](ContextMenuButton* btn)
			{
				OnItemClicked(std::any_cast<int>(btn->Tag));
				if (SupportMultiSelect)
				{
					// Don't hide in multi-select, so user can edit multiple elements instead of just one
					UpdateButtons();
					if (_popupMenu != nullptr)
					{
						_popupMenu->PerformLayout();
					}

				}
				else
				{
					if (_popupMenu != nullptr)
					{
						_popupMenu->Hide();
					}
				}
			});
		}

		// Check if menu hs been already shown
		if (_popupMenu->Visible)
		{
			if (!SupportMultiSelect)
				_popupMenu->Hide();
			return;
		}

		PopupShowing(this);

		// Check if has any items
		if (items.Count() > 0)
		{
			UpdateButtons();

			// Show dropdown list
			_popupMenu->MinimumWidth = Width;
			_popupMenu->Show(this, Float2(1, Height));

			// Adjust menu position if it is not the down direction
			if (_popupMenu->Direction == ContextMenuDirection::RightUp)
			{
				Float2 position = _popupMenu->RootWindow()->Window()->GetPosition();
				_popupMenu->RootWindow()->Window()->SetPosition(Float2(position.x, position.y - Height));
			}
		}
	}

	void ComboBox::OnLayoutMenuButton(ContextMenuButton* button, int index, bool construct)
	{
		button->SetChecked(_selectedIndices.Contains(index));
		if (tooltips.Count() > index)
			button->TooltipText = tooltips[index];
	}

	ContextMenu* ComboBox::OnCreatePopup()
	{
		if (PopupCreate.IsBinded())
			return PopupCreate(this);
		return New<ContextMenu>();
	}

	void ComboBox::UpdateButtons()
	{
		if (_popupMenu->Items.operator->().Count() != items.Count())
		{
			List<ContextMenuItem*> itemControls = _popupMenu->Items;
			for (auto e : itemControls)
			{
				e->Dispose();
			}
			if (Sorted)
			{
				Sorting::QuickSort(items);
			}
			for (int i = 0; i < items.Count(); i++)
			{
				ContextMenuButton* btn = _popupMenu->AddButton(items[i]);
				OnLayoutMenuButton(btn, i, true);
				btn->Tag = i;
			}
		}
		else
		{
			List<ContextMenuItem*> itemControls = _popupMenu->Items;
			if (Sorted)
			{
				Sorting::QuickSort(items);
			}
			for (int i = 0; i < items.Count(); i++)
			{
				ContextMenuButton* btn;
				if (TypeTryCast<ContextMenuButton>(itemControls[i], btn))
				{
					btn->Text = items[i];
					OnLayoutMenuButton(btn, i, true);
				}
			}
		}
	}

	void ComboBox::__SetSelectedIndex(int value)
	{
		// Clamp index
		value = Math::Min(value, items.Count() - 1);

		// Check if index will change
		if (value != SelectedIndex)
		{
			// Select
			_selectedIndices.Clear();
			if (value != -1)
				_selectedIndices.Add(value);
			OnSelectedIndexChanged();
		}
	}

	void ComboBox::__SetSelection(List<int>& value)
	{
		ENGINE_ASSERT(!(!SupportMultiSelect && value.Count() > 1))

		for (int i = 0; i < value.Count(); i++)
		{
			int index = value[i];
			if (index < 0 || index >= items.Count())
			{
				ENGINE_ASSERT(false);
			}
		}

		Function<bool(const int& a, const int& b)> func = [](const int& a, const int& b)
		{
			return a == b;
		};
		if (!ListExtensions::SequenceEqual(_selectedIndices, value, func))
		{
			// Select
			_selectedIndices.Clear();
			_selectedIndices.Add(value);
			OnSelectedIndexChanged();
		}
	}

	bool ComboBox::__GetIsPopupOpened()
	{
		return _popupMenu != nullptr && _popupMenu->IsOpened;
	}
} // SE

File: Src/Engine/Editor/GUI/ComboBox.cs
// Managed editor GUI feature implementation.
using System;
using System.Collections.Generic;
using SE.GUI;

namespace SE.Editor.GUI
{
    public class ComboBox : Control
    {
        public const float DefaultHeight = 18.0f;

        private readonly List<string> m_Items = new List<string>();
        private readonly List<string> m_Tooltips = new List<string>();
        private readonly List<int> m_SelectedIndices = new List<int>(4);
        private ContextMenu? m_PopupMenu;
        private bool m_IsPressed;

        public ComboBox()
            : this(0.0f, 0.0f)
        {
        }

        public ComboBox(float x, float y, float width = 120.0f)
            : base(new Rectangle(x, y, width, DefaultHeight))
        {
            MaximumItemsInViewCount = 20;
            BackgroundColor = Style.Current.BackgroundNormal;
        }

        public event Action<ComboBox>? SelectedIndexChanged;
        public event Action<ComboBox>? PopupShowing;

        public IList<string> Items => m_Items;
        public IList<string> Tooltips => m_Tooltips;
        public bool Sorted { get; set; }
        public bool SupportMultiSelect { get; set; }
        public int MaximumItemsInViewCount { get; set; }
        public Func<ComboBox, ContextMenu>? PopupCreate { get; set; }
        public ContextMenu? Popup => m_PopupMenu;
        public bool IsPopupOpened => m_PopupMenu?.IsOpened == true;
        public bool HasSelection => m_SelectedIndices.Count != 0;
        public string Text { get; private set; } = string.Empty;

        public string SelectedItem
        {
            get => m_SelectedIndices.Count == 1 ? m_Items[m_SelectedIndices[0]] : string.Empty;
            set => SelectedIndex = m_Items.IndexOf(value);
        }

        public int SelectedIndex
        {
            get => m_SelectedIndices.Count == 1 ? m_SelectedIndices[0] : -1;
            set
            {
                int clamped = m_Items.Count == 0 ? -1 : Math.Min(Math.Max(value, -1), m_Items.Count - 1);
                if (SelectedIndex == clamped)
                    return;

                m_SelectedIndices.Clear();
                if (clamped != -1)
                    m_SelectedIndices.Add(clamped);
                OnSelectedIndexChanged();
            }
        }

        public IReadOnlyList<int> Selection => m_SelectedIndices;

        public void SetSelection(IEnumerable<int> selection)
        {
            List<int> next = new List<int>();
            foreach (int index in selection)
            {
                if (index < 0 || index >= m_Items.Count)
                    throw new ArgumentOutOfRangeException(nameof(selection), "Selection index is outside the item range.");
                next.Add(index);
            }

            if (!SupportMultiSelect && next.Count > 1)
                throw new InvalidOperationException("ComboBox does not support multiple selected items.");

            if (SequenceEqual(m_SelectedIndices, next))
                return;

            m_SelectedIndices.Clear();
            m_SelectedIndices.AddRange(next);
            OnSelectedIndexChanged();
        }

        public void ClearItems()
        {
            SelectedIndex = -1;
            m_Items.Clear();
            RefreshText();
        }

        public void AddItem(string item)
        {
            m_Items.Add(item);
            RefreshText();
        }

        public void AddItems(IEnumerable<string> items)
        {
            m_Items.AddRange(items);
            RefreshText();
        }

        public void SetItems(IEnumerable<string> items)
        {
            SelectedIndex = -1;
            m_Items.Clear();
            m_Items.AddRange(items);
            RefreshText();
        }

        public bool IsSelected(string item)
        {
            return IsSelected(m_Items.IndexOf(item));
        }

        public bool IsSelected(int index)
        {
            return index >= 0 && m_SelectedIndices.Contains(index);
        }

        public override bool OnMouseDown(Float2 location, MouseButton button)
        {
            if (button != MouseButton.Left)
                return false;

            m_IsPressed = true;
            Root?.StartTrackingMouse(this);
            return true;
        }

        public override bool OnMouseUp(Float2 location, MouseButton button)
        {
            if (button != MouseButton.Left || !m_IsPressed)
                return false;

            m_IsPressed = false;
            Root?.EndTrackingMouse();
            if (location.X >= 0.0f && location.Y >= 0.0f && location.X <= Width && location.Y <= Height)
                ShowPopup();
            return true;
        }

        public override void ClearState()
        {
            m_IsPressed = false;
            base.ClearState();
        }

        protected override void OnDraw()
        {
            base.OnDraw();

            Font? font = Style.Current.FontMedium;
            if (ReferenceEquals(font, null) || string.IsNullOrEmpty(Text))
                return;

            Rectangle bounds = ScreenBounds;
            Color color = EnabledInHierarchy ? Style.Current.Foreground : Style.Current.ForegroundDisabled;
            Render2D.RenderText(font, Text, ref bounds, ref color, TextAlignment.Near, TextAlignment.Center, TextWrapping.NoWrap);
        }

        public void ShowPopup()
        {
            m_PopupMenu ??= OnCreatePopup();
            m_PopupMenu.MaximumItemsInViewCount = MaximumItemsInViewCount;

            if (m_PopupMenu.IsOpened)
            {
                if (!SupportMultiSelect)
                    m_PopupMenu.Hide();
                return;
            }

            PopupShowing?.Invoke(this);
            if (m_Items.Count == 0)
                return;

            UpdateButtons();
            m_PopupMenu.MinimumWidth = Width;
            m_PopupMenu.Show(this, 1.0f, Height);
        }

        protected virtual void OnSelectedIndexChanged()
        {
            TooltipText = m_Tooltips.Count == m_Items.Count && m_SelectedIndices.Count == 1
                ? m_Tooltips[m_SelectedIndices[0]]
                : string.Empty;
            RefreshText();
            SelectedIndexChanged?.Invoke(this);
        }

        protected virtual void OnItemClicked(int index)
        {
            if (SupportMultiSelect)
            {
                if (m_SelectedIndices.Contains(index))
                    m_SelectedIndices.Remove(index);
                else
                    m_SelectedIndices.Add(index);
                OnSelectedIndexChanged();
                UpdateButtons();
            }
            else
            {
                SelectedIndex = index;
                m_PopupMenu?.Hide();
            }
        }

        protected virtual void OnLayoutMenuButton(ContextMenuButton button, int index, bool construct)
        {
            button.SetChecked(m_SelectedIndices.Contains(index));
            if (m_Tooltips.Count > index)
                button.TooltipText = m_Tooltips[index];
        }

        protected virtual ContextMenu OnCreatePopup()
        {
            return PopupCreate != null ? PopupCreate(this) : new ContextMenu();
        }

        private void UpdateButtons()
        {
            if (m_PopupMenu == null)
                return;

            m_PopupMenu.DisposeAllItems();
            if (Sorted)
                m_Items.Sort(StringComparer.OrdinalIgnoreCase);

            for (int i = 0; i < m_Items.Count; i++)
            {
                int itemIndex = i;
                ContextMenuButton button = m_PopupMenu.AddButton(m_Items[i], _ => OnItemClicked(itemIndex));
                button.Tag = itemIndex;
                OnLayoutMenuButton(button, itemIndex, true);
            }
        }

        private void RefreshText()
        {
            Text = m_SelectedIndices.Count switch
            {
                0 => string.Empty,
                1 when m_SelectedIndices[0] >= 0 && m_SelectedIndices[0] < m_Items.Count => m_Items[m_SelectedIndices[0]],
                1 => string.Empty,
                _ => "Multiple Values",
            };
        }

        private static bool SequenceEqual(IReadOnlyList<int> left, IReadOnlyList<int> right)
        {
            if (left.Count != right.Count)
                return false;

            for (int i = 0; i < left.Count; i++)
            {
                if (left[i] != right[i])
                    return false;
            }

            return true;
        }
    }
}
